src/Interpreters/MergeJoin.cpp

```diff
@@ -532,16 +532,31 @@ MergeJoin::MergeJoin(std::shared_ptr<TableJoin> table_join_, const Block & right
     addConditionJoinColumn(right_sample_block, JoinTableSide::Right);
     JoinCommon::splitAdditionalColumns(key_names_right, right_sample_block, right_table_keys, right_columns_to_add);
 
+    const NameSet required_right_keys = table_join->requiredRightKeys();
     for (const auto & right_key : key_names_right)
     {
+        if (table_join->hasUsing() && !required_right_keys.contains(right_key))
```

Comment:
It's quite unclear why it checks for `LowCardinality`

Reply:
We convert every lowcardinality key column to full inside MergeJoin. But in order to have the same types in output we restore lowcardinality back. For that we save such key names in `lowcard_right_keys`. The issue is that to case described below in a comment we need to have non-lowcardinality type in output, but return lowcard.